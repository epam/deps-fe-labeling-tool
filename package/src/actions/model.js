
import { batch } from 'react-redux'
import { createAction } from 'redux-actions'
import { LabelType, LABEL_TYPE_NAME } from '@/models/Label'
import { Markup } from '@/models/Markup'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector, fieldsToDeleteSelector } from '@/selectors/model'
import {
  updateLabelsWithSettings,
  updateTablesWithSettings
} from './markup'

const FEATURE_NAME = 'MODEL'

const storeFields = createAction(
  `${FEATURE_NAME}/STORE`
)

const storeInitialFields = createAction(
  `${FEATURE_NAME}/STORE_INITIAL_FIELDS`
)

const updateInitialFields = createAction(
  `${FEATURE_NAME}/UPDATE_INITIAL_FIELDS`
)

const enableDeleteFieldsMode = createAction(
  `${FEATURE_NAME}/ENABLE_DELETE_FIELD_MODE`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

const updateFieldsToDelete = createAction(
  `${FEATURE_NAME}/UPDATE_FIELDS_TO_DELETE`
)

const cancelDeleteFieldsMode = createAction(
  `${FEATURE_NAME}/CANCEL_DELETE_FIELDS_MODE`
)

const confirmFieldsDeletion = () => (dispatch, getState) => {
  const state = getState()
  const markup = markupSelector(state)
  const fields = fieldsSelector(state)
  const fieldsToDelete = fieldsToDeleteSelector(state)
  const newFields = fields.filter((field) => fieldsToDelete.every((f) => f.code !== field.code))
  const assignedObjects = markup
    ? Markup.getAllObjects(markup).filter((markupObject) => fieldsToDelete.find((field) => field.code === markupObject.fieldCode))
    : []

  batch(() => {
    assignedObjects.forEach((markupObject) => {
      if (markupObject.typeName === LABEL_TYPE_NAME) {
        dispatch(updateLabelsWithSettings({
          ...markupObject,
          fieldCode: '',
          type: LabelType.UNASSIGNED,
          index: undefined
        }))
      } else {
        dispatch(updateTablesWithSettings({
          ...markupObject,
          fieldCode: '',
          index: undefined
        }))
      }
    })
    dispatch(storeFields(newFields))
    dispatch(cancelDeleteFieldsMode())
  })
}

export {
  enableDeleteFieldsMode,
  resetDefault,
  storeFields,
  storeInitialFields,
  updateFieldsToDelete,
  updateInitialFields,
  confirmFieldsDeletion,
  cancelDeleteFieldsMode
}
