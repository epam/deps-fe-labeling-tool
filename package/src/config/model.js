import { FieldType } from '@/enums/FieldType'
import { Field } from '@/models/Field'
import { PairFieldMeta, ListFieldMeta, EnumFieldMeta } from '@/models/FieldMeta'

const fields = [
  new Field(
    'dates',
    'Dates',
    FieldType.DATE
  ),
  new Field(
    'firstName',
    'First Name',
    FieldType.STRING
  ),
  new Field(
    'lastName',
    'Last Name',
    FieldType.STRING,
    undefined,
    true
  ),
  new Field(
    'companyName',
    'Company Name',
    FieldType.STRING,
    undefined,
    true
  ),
  new Field(
    'city',
    'City',
    FieldType.ENUM,
    new EnumFieldMeta([
      'Anytown',
      'Sometown',
      'San Diego',
      'Los Angeles'
    ]),
    true
  ),
  new Field(
    'countryStates',
    'Country States',
    FieldType.LIST,
    new ListFieldMeta(FieldType.STRING),
    true
  ),
  new Field(
    'dateMatching',
    'Date Matching',
    FieldType.LIST,
    new ListFieldMeta(FieldType.CHECKMARK),
    true
  ),
  new Field(
    'date',
    'Date',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING)
  ),
  new Field(
    'zipCodes',
    'Zip Codes',
    FieldType.LIST,
    new ListFieldMeta(
      FieldType.PAIR,
      new PairFieldMeta(FieldType.STRING, FieldType.STRING)
    ),
    true
  ),
  new Field(
    'homeAddress',
    'Home Address',
    FieldType.STRING
  ),
  new Field(
    'businessAddress',
    'Business Address',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING),
    true
  ),
  new Field(
    'taxpayerSpousesInfo',
    'Taxpayer Spouse\'s Info',
    FieldType.TABLE
  ),
  new Field(
    'taxpayerInfo',
    'Taxpayer Info',
    FieldType.TABLE,
    undefined,
    true
  ),
  new Field(
    'tables',
    'Tables',
    FieldType.LIST,
    new ListFieldMeta(FieldType.TABLE)
  ),
  new Field(
    'uniqueIdentifier',
    'Unique Identifier',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING)
  ),
  new Field(
    'signatory',
    'Signatory',
    FieldType.CHECKMARK
  ),
  new Field(
    'returnTranscript',
    'Return Transcript',
    FieldType.CHECKMARK
  ),
  new Field(
    'formNumbers',
    'Form Numbers',
    FieldType.LIST,
    new ListFieldMeta(
      FieldType.ENUM,
      new EnumFieldMeta([
        '1098',
        '5498',
        '1040'
      ])
    )
  )
]

export {
  fields
}
