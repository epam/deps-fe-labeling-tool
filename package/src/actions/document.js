import { createAction } from 'redux-actions'

const FEATURE_NAME = 'DOCUMENT'

const storeDocument = createAction(`${FEATURE_NAME}/STORE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  storeDocument
}
