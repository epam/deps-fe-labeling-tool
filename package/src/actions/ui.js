import { createAction } from 'redux-actions'

const FEATURE_NAME = 'UI'

const setMarkupObjectsFilter = createAction(
  `${FEATURE_NAME}/SET_MARKUP_OBJECTS_FILTER`
)

const setAssignToFieldsFilter = createAction(
  `${FEATURE_NAME}/SET_ASSIGN_TO_FIELDS_FILTER`
)

const setExpandedListKeys = createAction(`${FEATURE_NAME}/SET_KEYS`)

const toggleExpandedListKey = createAction(`${FEATURE_NAME}/TOGGLE_KEY`)
const setActiveSidebar = createAction(`${FEATURE_NAME}/SET_ACTIVE_SIDEBAR`)

const addTemporaryFieldIndex = createAction(`${FEATURE_NAME}/ADD_TEMPORARY_FIELD`)
const deleteTemporaryFieldIndex = createAction(`${FEATURE_NAME}/DELETE_TEMPORARY_FIELD_INDEX`)

const resetDefault = createAction(`${FEATURE_NAME}/RESET_DEFAULT`)

export {
  setMarkupObjectsFilter,
  setActiveSidebar,
  setExpandedListKeys,
  toggleExpandedListKey,
  addTemporaryFieldIndex,
  setAssignToFieldsFilter,
  deleteTemporaryFieldIndex,
  resetDefault
}
