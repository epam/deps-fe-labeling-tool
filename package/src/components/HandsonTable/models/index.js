
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
