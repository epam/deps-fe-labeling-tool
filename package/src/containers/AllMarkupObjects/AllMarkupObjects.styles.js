import styled, { css } from 'styled-components'
import { Button } from '@/components/Button'
import { Collapse } from '@/components/Collapse'

const ButtonIcon = styled(Button.Icon)`
  display: flex;
  align-items: center;
`

const DocumentName = styled.h1`
  min-height: 3.2rem;
  display: flex;
  align-items: center;
  margin: 0;
`

const ExtraIcons = styled.div`
  display: flex;
`

const FiltersWrapper = styled.div`
  display: flex;
  margin: 0.8rem 0;
`

const MarkupCollapse = styled(Collapse)`
  width: 29.2rem;
  align-items: center;
  border-left: 0.2rem solid transparent;
  padding: 1rem 1.6rem !important;
  cursor: pointer;
  ${(props) => props.$unassigned && css`
      border-left: 0.2rem solid red;
      margin: 0.2rem 0;
    `}

  & .ant-collapse-content-box {
    padding: 0 !important;
  }
`

export {
  ButtonIcon,
  FiltersWrapper,
  DocumentName,
  ExtraIcons,
  MarkupCollapse as Collapse
}
