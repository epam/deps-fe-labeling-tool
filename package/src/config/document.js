import { pages } from '@/assets/document'
import { KnownLanguage } from '@/enums/KnownLanguage'
import { Document } from '@/models/Document'

const documentForLabeling = new Document(
  pages,
  KnownLanguage.ENGLISH,
  'Demo',
  'Tesseract'
)

export {
  documentForLabeling
}
