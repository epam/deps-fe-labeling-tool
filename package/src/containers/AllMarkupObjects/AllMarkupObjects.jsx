import React, {
  useMemo,
  useCallback,
  useState
} from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { setMarkupObjectsFilter, setExpandedListKeys } from '@/actions/ui'
import { CollapseIcon } from '@/components/Icons/CollapseIcon'
import { ExpandIcon } from '@/components/Icons/ExpandIcon'
import { Tooltip } from '@/components/Tooltip'
import {
  Group,
  FieldsViewSwitcher
} from '@/containers/FieldsViewSwitcher'
import { MarkupObjects } from '@/containers/MarkupObjects'
import { SearchInput } from '@/containers/SearchInput'
import { FieldType } from '@/enums/FieldType'
import { fieldShape } from '@/models/Field'
import { documentNameSelector, extraNameSelector } from '@/selectors/document'
import { fieldsSelector } from '@/selectors/model'
import { markupObjectsFilterSelector, expandedListKeysSelector } from '@/selectors/ui'
import {
  ButtonIcon,
  DocumentName,
  ExtraIcons,
  FiltersWrapper,
  Collapse
} from './AllMarkupObjects.styles'

const RESOURCE_EXPAND_ALL_BUTTON_TEXT = 'Expand all'
const RESOURCE_COLLAPSE_ALL_BUTTON_TEXT = 'Collapse all'
const RESOURCE_VIEW_ALL = 'View all'
const RESOURCE_FIELD_TYPE = 'View according to field type'

const GROUPING_TYPE = {
  VIEW_ALL: 'viewAll',
  FIELD_TYPE: 'fieldType'
}

const FIELDS_GROUPS = [
  new Group(GROUPING_TYPE.VIEW_ALL, RESOURCE_VIEW_ALL),
  new Group(GROUPING_TYPE.FIELD_TYPE, RESOURCE_FIELD_TYPE)
]

const AllMarkupObjects = ({
  documentName,
  extraName,
  fields,
  markupObjectsFilter,
  setMarkupObjectsFilter,
  expandedListKeys,
  setExpandedListKeys
}) => {
  const [fieldsGrouping, setFieldsGrouping] = useState(GROUPING_TYPE.VIEW_ALL)
  const [activeDocument, setActiveDocument] = useState(documentName)

  const onChangeFilter = useCallback(
    (value) => {
      setMarkupObjectsFilter(value)
    }, [setMarkupObjectsFilter]
  )

  const allListKeys = useMemo(() => fields
    .filter(({ fieldType }) => fieldType === FieldType.LIST)
    .map((field) => field.code)
  , [fields])

  const areAllListsExpanded = useMemo(() => {
    return (
      allListKeys.length === expandedListKeys.length &&
      allListKeys.every((key) => expandedListKeys.includes(key))
    )
  }, [expandedListKeys, allListKeys])

  const collapseAll = useCallback((e) => {
    e.stopPropagation()
    setExpandedListKeys([])
  }, [setExpandedListKeys])

  const expandAll = useCallback((e) => {
    e.stopPropagation()
    setExpandedListKeys(allListKeys)
  }, [allListKeys, setExpandedListKeys])

  const isExtraVisible = useMemo(() => (
    !!activeDocument && !!allListKeys.length
  ), [activeDocument, allListKeys])

  const toggleActiveDocument = useCallback((e) => {
    setActiveDocument((activeDocument) => activeDocument ? '' : documentName)
  }, [documentName])

  const renderExtra = useMemo(() => (
    isExtraVisible && (
      <ExtraIcons>
        <Tooltip title={RESOURCE_COLLAPSE_ALL_BUTTON_TEXT}>
          <ButtonIcon
            disabled={!expandedListKeys.length}
            onClick={collapseAll}
            icon={<CollapseIcon />}
          >
          </ButtonIcon>
        </Tooltip>
        <Tooltip title={RESOURCE_EXPAND_ALL_BUTTON_TEXT}>
          <ButtonIcon
            disabled={areAllListsExpanded}
            icon={<ExpandIcon />}
            onClick={expandAll}
          >
          </ButtonIcon>
        </Tooltip>
      </ExtraIcons>
    )
  ), [
    areAllListsExpanded,
    collapseAll,
    expandAll,
    expandedListKeys.length,
    isExtraVisible
  ])

  return (
    <Collapse
      activeKey={activeDocument}
      collapseId={documentName}
      onChange={toggleActiveDocument}
      header={
        <DocumentName>
          {`${extraName} | ${documentName}`}
        </DocumentName>
      }
      extra={renderExtra}
    >
      <FiltersWrapper>
        <SearchInput
          value={markupObjectsFilter}
          onChange={onChangeFilter}
        />
        <FieldsViewSwitcher
          groups={FIELDS_GROUPS}
          setFieldsGrouping={setFieldsGrouping}
        />
      </FiltersWrapper>
      <MarkupObjects
        fieldsGrouping={fieldsGrouping}
        setFilter={onChangeFilter}
      />
    </Collapse>
  )
}

AllMarkupObjects.propTypes = {
  documentName: PropTypes.string,
  extraName: PropTypes.string,
  fields: PropTypes.arrayOf(fieldShape),
  markupObjectsFilter: PropTypes.string.isRequired,
  setMarkupObjectsFilter: PropTypes.func.isRequired,
  expandedListKeys: PropTypes.arrayOf(PropTypes.string),
  setExpandedListKeys: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  documentName: documentNameSelector(state),
  extraName: extraNameSelector(state),
  fields: fieldsSelector(state),
  markupObjectsFilter: markupObjectsFilterSelector(state),
  expandedListKeys: expandedListKeysSelector(state)
})

const mapDispatchToProps = {
  setMarkupObjectsFilter,
  setExpandedListKeys
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(AllMarkupObjects)

export {
  ConnectedComponent as AllMarkupObjects
}
