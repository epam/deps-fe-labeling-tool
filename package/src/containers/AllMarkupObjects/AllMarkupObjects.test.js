import { mockUiActions } from '@/mocks/actions/ui'
import { mockComponent } from '@/mocks/mockComponent'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockUuid } from '@/mocks/mockUuid'
import { mockDocumentSelectors } from '@/mocks/selectors/document'
import { mockModelSelectors } from '@/mocks/selectors/model'
import { mockUiSelectors } from '@/mocks/selectors/ui'
import React from 'react'
import { shallow } from 'enzyme'
import { setMarkupObjectsFilter, setExpandedListKeys } from '@/actions/ui'
import { FieldType } from '@/enums/FieldType'
import { Field } from '@/models/Field'
import { LABEL_TYPE_NAME } from '@/models/Label'
import { AllMarkupObjects } from './AllMarkupObjects'
import { ButtonIcon } from './AllMarkupObjects.styles'

jest.mock('uuid', () => mockUuid)
jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/actions/ui', () => mockUiActions)
jest.mock('@/selectors/model', () => mockModelSelectors)
jest.mock('@/selectors/document', () => mockDocumentSelectors)
jest.mock('@/selectors/ui', () => mockUiSelectors)
jest.mock('@/containers/MarkupObjects', () => mockComponent('MarkupObjects'))

const MOCK_FIELD_NAME = 'MOCK_FIELD_1'

const { WrappedComponent, mapStateToProps, mapDispatchToProps } = AllMarkupObjects

describe('Container: AllMarkupObject', () => {
  describe('mapStateToProps', () => {
    it('should pass state as props to WrappedComponent', () => {
      const { props } = mapStateToProps()
      const expectedProps = {
        documentName: mockDocumentSelectors.documentNameSelector(),
        extraName: mockDocumentSelectors.extraNameSelector(),
        fields: mockModelSelectors.fieldsSelector(),
        markupObjectsFilter: mockUiSelectors.markupObjectsFilterSelector(),
        expandedListKeys: mockUiSelectors.expandedListKeysSelector()
      }

      expect(props).toEqual(expectedProps)
    })
  })

  describe('mapDispatchToProps', () => {
    const { props } = mapDispatchToProps()

    it('should call for setMarkupObjectsFilter action from props', () => {
      props.setMarkupObjectsFilter()
      expect(setMarkupObjectsFilter).toHaveBeenCalledTimes(1)
    })

    it('should pass setExpandedListKeys action as setExpandedListKeys prop', () => {
      const { props } = mapDispatchToProps()
      props.setExpandedListKeys()
      expect(setExpandedListKeys).toHaveBeenCalledTimes(1)
    })
  })

  describe('ConnectedComponent', () => {
    let defaultProps
    let wrapper
    beforeEach(() => {
      defaultProps = {
        documentName: mockDocumentSelectors.documentNameSelector(),
        extraName: mockDocumentSelectors.extraNameSelector(),
        fields: [new Field(MOCK_FIELD_NAME, LABEL_TYPE_NAME, FieldType.LIST)],
        markupObjectsFilter: mockUiSelectors.markupObjectsFilterSelector(),
        expandedListKeys: mockUiSelectors.expandedListKeysSelector(),
        ...mapDispatchToProps().props
      }

      wrapper = shallow(<WrappedComponent {...defaultProps} />)
    })

    it('should render correct layout', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call setExpandedListKeys with correct arg when calling collapse all button onClick', () => {
      const mockClickEvent = {
        stopPropagation: jest.fn()
      }

      const collapseAll = shallow(<div>
        {
          wrapper.props().extra
        }
      </div>).find(ButtonIcon).first()
      const collapseAllProps = collapseAll.find(ButtonIcon).props()
      collapseAllProps.onClick(mockClickEvent)
      expect(setExpandedListKeys).nthCalledWith(1, [])
    })

    it('should call setExpandedListKeys with correct arg when calling expand all button onClick', () => {
      const mockClickEvent = {
        stopPropagation: jest.fn()
      }
      const expandAll = shallow(<div>
        {
          wrapper.props().extra
        }
      </div>).find(ButtonIcon).at(1)
      const expandAllProps = expandAll.find(ButtonIcon).props()
      expandAllProps.onClick(mockClickEvent)
      expect(setExpandedListKeys).nthCalledWith(
        1,
        defaultProps.fields
          .filter(({ fieldType }) => fieldType === FieldType.LIST)
          .map((field) => field.code)
      )
    })
  })
})
