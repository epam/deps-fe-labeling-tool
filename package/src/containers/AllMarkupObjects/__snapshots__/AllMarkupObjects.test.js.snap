// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: AllMarkupObject ConnectedComponent should render correct layout 1`] = `
<AllMarkupObjectsstyles__MarkupCollapse
  activeKey="Mock Title"
  collapseId="Mock Title"
  extra={
    <AllMarkupObjectsstyles__ExtraIcons>
      <Tooltip
        title="Collapse all"
      >
        <AllMarkupObjectsstyles__ButtonIcon
          disabled={false}
          icon={<CollapseIcon />}
          onClick={[Function]}
        />
      </Tooltip>
      <Tooltip
        title="Expand all"
      >
        <AllMarkupObjectsstyles__ButtonIcon
          disabled={false}
          icon={<ExpandIcon />}
          onClick={[Function]}
        />
      </Tooltip>
    </AllMarkupObjectsstyles__ExtraIcons>
  }
  header={
    <AllMarkupObjectsstyles__DocumentName>
      Mock Extra Name | Mock Title
    </AllMarkupObjectsstyles__DocumentName>
  }
  onChange={[Function]}
>
  <AllMarkupObjectsstyles__FiltersWrapper>
    <SearchInput
      onChange={[Function]}
      value="testMarkupFilter"
    />
    <FieldsViewSwitcher
      groups={
        Array [
          Group {
            "code": "viewAll",
            "name": "View all",
          },
          Group {
            "code": "fieldType",
            "name": "View according to field type",
          },
        ]
      }
      setFieldsGrouping={[Function]}
    />
  </AllMarkupObjectsstyles__FiltersWrapper>
  <MarkupObjects
    fieldsGrouping="viewAll"
    setFilter={[Function]}
  />
</AllMarkupObjectsstyles__MarkupCollapse>
`;
